File: v1-tote-ops/screenshot_demo.py
"""Capture every Tote-Ops screen for the v1 (before) asset set.

DEMO ONLY. Not part of the production Tote-Ops application.

Assumes the app is already running (see docs/plan.md) and that seed_demo.py
has been run against the same database.

Writes to ../assets/v1-tote-ops/<width>/<name>.png

Run:  python screenshot_demo.py
"""

import os
import sys
from pathlib import Path

from playwright.sync_api import sync_playwright
from sqlalchemy import func

from app.config import settings
from app.database import _SessionFactory
from app.models.gradeout import Gradeout
from app.models.invoice import Invoice

BASE_URL = os.environ.get("DEMO_BASE_URL", "http://127.0.0.1:8011")
OUT_ROOT = Path(__file__).resolve().parent.parent / "assets" / "v1-tote-ops"

# The app is iPad-primary, so 768 is the design target and 1280 shows the
# desktop layout. Both widths are captured for every screen.
VIEWPORTS = {"desktop-1280": 1280, "ipad-768": 768}


def pick_targets() -> dict:
    """Choose concrete record IDs so the detail screens show a busy record
    rather than an arbitrary one."""
    db = _SessionFactory()
    try:
        busiest = (
            db.query(Gradeout.supplier_id, func.count(Gradeout.id).label("n"))
            .group_by(Gradeout.supplier_id)
            .order_by(func.count(Gradeout.id).desc())
            .first()
        )
        latest_gradeout = (
            db.query(Gradeout)
            .order_by(Gradeout.date_received.desc(), Gradeout.created_at.desc())
            .first()
        )
        latest_invoice = db.query(Invoice).order_by(Invoice.month.desc()).first()
        if not (busiest and latest_gradeout and latest_invoice):
            sys.exit("Database looks unseeded — run seed_demo.py first.")
        return {
            "supplier_id": busiest[0],
            "gradeout_id": latest_gradeout.id,
            "invoice_month": latest_invoice.month.strftime("%Y-%m"),
        }
    finally:
        db.close()


def routes(t: dict) -> list[tuple[str, str]]:
    """(filename, path) pairs, in the order a visitor would walk the app."""
    return [
        ("01-dashboard", "/?period=month"),
        ("01b-dashboard-year", "/?period=year"),
        ("02-suppliers-list", "/suppliers"),
        ("03-supplier-detail", f"/suppliers/{t['supplier_id']}"),
        ("04-suppliers-filtered-pickup-confirmed", "/suppliers?action_stage=pickup_confirmed"),
        ("05-gradeouts-list", "/gradeouts"),
        ("06-gradeout-new", "/gradeouts/new"),
        ("07-gradeout-edit", f"/gradeouts/{t['gradeout_id']}/edit"),
        ("08-leads-list", "/leads"),
        ("09-leads-filtered-contacted", "/leads?status=contacted"),
        ("10-invoices-list", "/invoices"),
        ("11-growth-calculator", "/growth"),
    ]


# The dashboard ships with its three panels collapsed. Screenshotting it
# as-is produces three empty cards, so they are opened before capture.
DASHBOARD_PANELS = ["#chart-toggle-btn", "#gradeouts-toggle-btn", "#followup-toggle-btn"]


def _expand_dashboard(page) -> None:
    for selector in DASHBOARD_PANELS:
        page.click(selector)
        page.wait_for_timeout(250)
    # Chart.js draws into a canvas only once its panel is visible.
    page.wait_for_timeout(1200)


def capture(page, label: str, path: str, out_dir: Path, after_load=None) -> None:
    resp = page.goto(f"{BASE_URL}{path}", wait_until="networkidle")
    if resp is None or resp.status >= 400:
        status = resp.status if resp else "no response"
        raise RuntimeError(f"{path} returned {status} — refusing to ship a broken screenshot")
    page.wait_for_timeout(600)
    if after_load:
        after_load(page)
    target = out_dir / f"{label}.png"
    page.screenshot(path=str(target), full_page=True)
    print(f"  {target.relative_to(OUT_ROOT.parent.parent)}")


def main() -> None:
    targets = pick_targets()
    with sync_playwright() as pw:
        browser = pw.chromium.launch()
        for vp_name, width in VIEWPORTS.items():
            out_dir = OUT_ROOT / vp_name
            out_dir.mkdir(parents=True, exist_ok=True)
            print(f"\n{vp_name}:")
            ctx = browser.new_context(
                viewport={"width": width, "height": 1000},
                device_scale_factor=2,
            )
            page = ctx.new_page()

            # Logged-out login screen, captured before authenticating.
            capture(page, "00-login", "/login", out_dir)

            page.goto(f"{BASE_URL}/login")
            page.fill('input[name="password"]', settings.APP_PASSWORD)
            page.click('button[type="submit"], input[type="submit"]')
            page.wait_for_url(f"{BASE_URL}/", timeout=15000)

            for label, path in routes(targets):
                hook = _expand_dashboard if label.startswith("01") else None
                capture(page, label, path, out_dir, after_load=hook)

            ctx.close()
        browser.close()
    print(f"\nWrote screenshots to {OUT_ROOT}")


if __name__ == "__main__":
    main()

File: v2-totetrack/screenshot_demo.mjs
/**
 * DEMO ONLY — captures every ToteTrack screen for the v2 (after) asset set.
 *
 * Point it at the built static export (served locally, or the live GitHub
 * Pages URL) and it walks every route at both viewport widths.
 *
 * Usage:
 *   node screenshot_demo.mjs                          # http://127.0.0.1:8022/tote-platform
 *   DEMO_BASE_URL=https://... node screenshot_demo.mjs
 */

import { chromium } from 'playwright'
import { mkdir } from 'node:fs/promises'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'

const HERE = dirname(fileURLToPath(import.meta.url))
const BASE_URL = process.env.DEMO_BASE_URL ?? 'http://127.0.0.1:8022/tote-platform'
const OUT_ROOT = join(HERE, '..', 'assets', 'v2-totetrack')

// Matches the widths used for the v1 capture so the before/after pairs line up.
const VIEWPORTS = { 'desktop-1280': 1280, 'ipad-768': 768 }

// `select` names screens whose right-hand pane only fills once a row is
// chosen. Capturing them untouched ships a screenshot of an empty state.
const ROUTES = [
  ['00-login', '/login/'],
  ['01-dashboard', '/dashboard/'],
  ['02-customers', '/customers/', { select: 'Select a customer to view details.' }],
  ['03-orders', '/orders/', { select: 'Select an order to view details.' }],
  ['04-leads', '/leads/', { select: 'Select a lead to view details.' }],
  ['05-invoices', '/invoices/'],
  ['06-calendar', '/calendar/'],
  // Support's right pane defaults to the new-ticket form rather than an empty
  // state, so it is never blank — but the ticket detail is the better shot.
  ['07-support', '/support/', { select: 'Select an issue to view details.' }],
]

/**
 * Clicks the first row of a master/detail screen and verifies the detail pane
 * actually filled. Throws rather than shipping a screenshot of an empty state.
 */
async function selectFirstRow(page, emptyStateText) {
  // Customers, leads and tickets render rows as buttons in a list; orders
  // render them as clickable table rows.
  const row = page.locator('ul li button, ul li a, tbody tr[tabindex]').first()
  if (!(await row.count())) throw new Error('no selectable rows found')
  await row.click()
  await page.waitForTimeout(1500)
  if (await page.getByText(emptyStateText, { exact: false }).count()) {
    throw new Error(`detail pane still empty after selecting a row (${emptyStateText})`)
  }
}

// The dashboard's revenue chart ships collapsed. Screenshotting it as-is
// leaves an empty card where the twelve-month trend should be.
async function expandRevenueChart(page) {
  const toggle = page.locator('button[aria-controls="revenue-chart-body"]')
  if ((await toggle.count()) && (await toggle.getAttribute('aria-expanded')) === 'false') {
    await toggle.click()
    // The collapse transition, then Recharts' own entrance animation, then a
    // margin — the series paints last and a short wait catches an empty plot.
    await page.waitForTimeout(3000)
  }
}

async function capture(page, label, path, outDir, afterLoad) {
  const resp = await page.goto(`${BASE_URL}${path}`, { waitUntil: 'networkidle' })
  if (!resp || resp.status() >= 400) {
    throw new Error(`${path} returned ${resp ? resp.status() : 'no response'} — refusing to ship a broken screenshot`)
  }
  // Framer Motion entrance animations and Recharts both settle after paint.
  await page.waitForTimeout(1600)
  if (afterLoad) await afterLoad(page)

  // `fullPage: true` resizes the viewport at capture time, which makes
  // Recharts' ResponsiveContainer re-measure and replay its entrance
  // animation — the screenshot then lands on an empty plot area. Resizing
  // first, settling, and then capturing the (now fully visible) viewport
  // avoids the race without touching the app's animation code.
  const height = await page.evaluate(() => document.body.scrollHeight)
  await page.setViewportSize({ width: page.viewportSize().width, height })
  await page.waitForTimeout(1800)

  const target = join(outDir, `${label}.png`)
  await page.screenshot({ path: target })
  await page.setViewportSize({ width: page.viewportSize().width, height: 1000 })
  console.log(`  ${label}.png`)
}

const browser = await chromium.launch()
const consoleErrors = []

for (const [vpName, width] of Object.entries(VIEWPORTS)) {
  const outDir = join(OUT_ROOT, vpName)
  await mkdir(outDir, { recursive: true })
  console.log(`\n${vpName}:`)

  const ctx = await browser.newContext({
    viewport: { width, height: 1000 },
    deviceScaleFactor: 2,
  })
  const page = await ctx.newPage()
  page.on('console', (msg) => {
    if (msg.type() === 'error') consoleErrors.push(`[${vpName}] ${msg.text()}`)
  })

  for (const [label, path, opts] of ROUTES) {
    let hook
    if (label === '01-dashboard') hook = expandRevenueChart
    else if (opts?.select) hook = (p) => selectFirstRow(p, opts.select)
    await capture(page, label, path, outDir, hook)
  }
  await ctx.close()
}

await browser.close()

console.log(`\nWrote screenshots to ${OUT_ROOT}`)

// A demo that logs errors in the console is not shippable — surface them.
if (consoleErrors.length) {
  console.error(`\n${consoleErrors.length} console error(s):`)
  for (const e of [...new Set(consoleErrors)]) console.error(`  ${e}`)
  process.exit(1)
}
console.log('No console errors.')
